File: main/java/mod/iceandshadow3/util/IteratorConcat.java
package mod.iceandshadow3.util;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

/**
 * Implements a list of iterators whose outputs are seamlessly concatenated.
 */
public class IteratorConcat<T, It extends Iterator<T>> implements ISupplier<T>, IResetable {
	protected List<It> iters;
	protected Iterator<It> listiter;
	protected It currentiter;

	public IteratorConcat() {
		iters = new LinkedList<It>();
	}

	public boolean add(It newiter) {
		iters.add(newiter);
		reset();
		return true;
	}

	protected void assureValidIter() {
		while (!currentiter.hasNext())
			if (listiter.hasNext())
				currentiter = listiter.next();
			else {
				currentiter = null;
				break;
			}
	}

	@Override
	public boolean hasNext() {
		assureValidIter();
		return currentiter != null;
	}

	@Override
	public T next() {
		assureValidIter();
		return currentiter != null ? currentiter.next() : null;
	}

	/**
	 * Resets the collection as much as possible. Does NOT reset contained
	 * iterators.
	 */
	@Override
	public boolean reset() {
		listiter = iters.iterator();
		currentiter = listiter.next();
		return false; // Does not reset individual iterators.
	}
}

File: src/main/scala/mod/iceandshadow3/util/IteratorConcat.scala
package mod.iceandshadow3.util

/** A Scala iterator that wraps Java iterators whose outputs are seamlessly concatenated and transformed by a function.
  * Iterables.concat does exist, but Scala seems to dislike it.
  */
class IteratorConcat[In, Out](
  converter: In => Out,
  iters: java.util.Iterator[In] *
) extends Iterator[Out] {
  private val listiter = iters.toIterator
  private var currentiter = listiter.next

  override def hasNext: Boolean = {
    while (!currentiter.hasNext && listiter.hasNext) currentiter = listiter.next
    currentiter.hasNext
  }

  override def next: Out = {
    if(hasNext) converter(currentiter.next) else null.asInstanceOf[Out]
  }
}